### PR TITLE
Demystifying api

### PR DESCRIPTION
**Pull request by pm0kjp**

### CODE REVIEW

**Comment by leemc-data-ed** [comment]
https://liascript.github.io/course/?https://raw.githubusercontent.com/arcus/education_modules/demystifying-api/demystifying_apis/demystifying_apis.md#1

File: demystifying_apis/demystifying_apis.md [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+
+* [Temple University](https://guides.temple.edu/APIs)
+* [University of California, Berkeley](https://guides.lib.berkeley.edu/information-studies/apis)
+* [University of Virginia](https://guides.lib.virginia.edu/c.php?g=515084&p=10046079)
+* [University of California, San Diego](https://ucsd.libguides.com/data-statistics/apis)
+
+Some APIs (particularly those linked to social media applications) have changing rules relating to their use, due to monetization of data and data privacy.  Read more in [Shifting landscapes of social media data for research](https://www.timeshighereducation.com/campus/shifting-landscapes-social-media-data-research).
+
+The United States Office of the National Coordinator for Health Information Technology (ONC) is driving an effort to develop and standardize APIs related to electronic health information.  Read more about [their project](https://www.healthit.gov/topic/scientific-initiatives/accelerating-apis-scientific-discovery)!
```
Comment (leemc-data-ed) [inline demystifying_apis/demystifying_apis.md:309]:
```suggestion
If you're interested in reading more about APIs in the biomedical and healthcare spaces, [the United States Office of the National Coordinator for Health Information Technology (ONC) is driving an effort to develop and standardize APIs related to electronic health information](https://www.healthit.gov/topic/scientific-initiatives/accelerating-apis-scientific-discovery). 
```
This is just a suggestion to try to make the link text itself a bit more descriptive, but I don't have super strong feelings about it.

File: demystifying_apis/demystifying_apis.md [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+
+</div>
+
+### Constructing an API Call in a Programming Language: API-Specific Packages
+
+We said on the last page that sometimes, specific APIs are very popular, and people write software (like a special library or package) to streamline using them.  This is true of APIs like those belonging to PubMed, the US Census Bureau, REDCap, and many more.  People write libraries to streamline the use of these popular APIs so you don't have to write quite as many lines of code as you saw on the previous page.
+
+For example, `PyCap` is a Python library written to streamline access to the REDCap API.  Because "get all the records from a given database" is such a frequently used API request, it got its own functionality in `PyCap`.  Compare the code below with the code on the previous page.  It's much shorter and less intimidating!
+
+![Six lines of Python code which get all of the records from a REDCap database ](media/pycap.png)
```
Comment (leemc-data-ed) [inline demystifying_apis/demystifying_apis.md:242]:
```suggestion
![Six lines of Python code which get all of the records from a REDCap database.](media/pycap.png)
```

File: demystifying_apis/demystifying_apis.md [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+
+## Interacting with an API 
+
+How do you interact with an Application Programming Interface?  Well, first, someone (like the IT department at the New York Times or the developers who work behind the scenes at REDCap) needs to have written an API, a set of rules that allows you, or your computer system, to interrogate their server, which is always listening for HTTP messages. 
+
+You'll use your computer to send a message to the server asking for whatever you need from the API.  Then the server will send back some combination of a return code (such as 200, good, or 404, not found) and, if things went well, the data you requested in your message.
```
Comment (leemc-data-ed) [inline demystifying_apis/demystifying_apis.md:117]:
```suggestion
You'll use your computer to send a message to the server asking for whatever you need from the API.  Then the server will send back some combination of a return code (such as 200, "good", or 404, "not found") and, if things went well, the data you requested in your message.
```

File: demystifying_apis/demystifying_apis.md [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+
+Let's look at an example!
+
+NCBI, the National Center for Biotechnology Information, has an API called "Entrez" that you can use for a number of purposes, including to get PubMed article IDs that meet your conditions.  They provide the following example on [their website](https://www.ncbi.nlm.nih.gov/books/NBK25500/):
+
+> Example: Get the PubMed IDs (PMIDs) for articles about breast cancer published in Science in 2008
+>
+> `https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pubmed&term=science[journal]+AND+breast+cancer+AND+2008[pdat]`
+
+Note that there are some terms added in the URL, parameters like `db` and `term`.  There are clearly some rules about how `term` is structured, and if we wanted to learn more about this API we could read their documentation to learn what parameters we could include in the URL.
```
Comment (leemc-data-ed) [inline demystifying_apis/demystifying_apis.md:167]:
```suggestion
Note that there are some terms that are included in the URL, parameters like `db` and `term`.  There are clearly some rules about how `term` is structured, and if we wanted to learn more about this API we could read their documentation to learn what parameters we could include in the URL.
```

File: demystifying_apis/demystifying_apis.md [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+
+![A screen full of XML encoded text that includes tags like "IdList", "eSearchResult", "QueryTranslation", and more.](media/entrez_screenshot.png)
+
+Sometimes you don't need to fully automate something, you're happy just to look at the results in your browser.  Maybe you're just doing a one-off search that you don't need to repeat, or you're testing things out.  Often, when you're learning how to use a new API, it can be helpful to do a browser test like this to make sure you understand the syntax of how to use the API and are getting the results you expect. 
+
+<div class = "help">
+<b style="color: rgb(var(--color-highlight));">Troubleshooting help</b><br>
+
+Entrez, like many APIs, has a limit to what you can do anonymously.  If you try to run this a few times in a row, you'll get an error message that starts with `{"error":"API rate limit exceeded"`...
```
Comment (leemc-data-ed) [inline demystifying_apis/demystifying_apis.md:178]:
```suggestion
E-utilities, like many APIs, has a limit to what you can do anonymously.  If you try to run this a few times in a row, you'll get an error message that starts with `{"error":"API rate limit exceeded"`...
```

**Review by leemc-data-ed — CHANGES_REQUESTED** [review]
Just a few small things, but I think this is awesome!

File: demystifying_apis/demystifying_apis.md [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+- Explain how to interact with an API
+
+@end
+
+good_first_module: true
+collection: demystifying, infrastructure_and_technology
+coding_required: false
+
+@sets_you_up_for
+
```
[the remark this answers is not in the document]
Comment (pm0kjp) [reply demystifying_apis/demystifying_apis.md:35]: I agree!  Adding in a new commit.

**Comment by pm0kjp** [comment]
ok @leemc-data-ed I think we're square?  I added your suggestions and added a smidge of metadata that you recommended as well!

**Pull request merged**